modules/dnn/include/opencv2/dnn/dnn.hpp

```diff
@@ -1135,6 +1147,76 @@ CV__DNN_INLINE_NS_BEGIN
                                    const Scalar& mean = Scalar(), bool swapRB=false, bool crop=false,
                                    int ddepth=CV_32F);
 
+    /**
+     * @brief Enum of image processing mode.
+     * @see ImagePParam
+     */
+    enum ImageProcessMode
+    {
+        DNN_PMODE_CENTER = 0,    // !< Default
+        DNN_PMODE_LETTERBOX = 1, // !< Resize image to the desired size while preserving the aspect ratio of original image.
+    };
+
+    /** @brief Image processing parameters.
+     *
+     * It includes all possible image processing operations and corresponding parameters.
+     *
+     * @see blobFromImageParam
+     * @note
+     * The order and usage of `scalefactor`, `size`, `mean`, `swapRB`, `crop` and `ddepth` are consistent
+     * with the function of @ref blobFromImage.
+    */
+    struct CV_EXPORTS_W_SIMPLE ImagePParam
+    {
+        CV_WRAP ImagePParam();
+        CV_WRAP ImagePParam(const Scalar& scalefactor, const Size& size = Size(), const Scalar& mean = Scalar(),
+                            bool swapRB = false, bool crop = false, int ddepth = CV_32F,
+                            DataLayout datalayout = DNN_LAYOUT_NCHW, ImageProcessMode mode = DNN_PMODE_CENTER);
+
+        CV_PROP_RW Scalar scalefactor; //!< scalefactor multiplier for input image values.
+        CV_PROP_RW Size size;    //!< Spatial size for output image.
+        CV_PROP_RW Scalar mean;  //!< Scalar with mean values which are subtracted from channels.
+        CV_PROP_RW bool swapRB;  //!< Flag which indicates that swap first and last channels
+        CV_PROP_RW bool crop;    //!< Flag which indicates whether image will be cropped after resize or not
+        CV_PROP_RW int ddepth;   //!< Depth of output blob. Choose CV_32F, CV_16F or CV_8U.
+        CV_PROP_RW DataLayout datalayout; //!< Order of output dimensions. Choose DNN_LAYOUT_NCHW or DNN_LAYOUT_NHWC.
+        CV_PROP_RW ImageProcessMode mode;   //!< Image processing mode. Choose DNN_PMODE_CENTER or DNN_PMODE_LETTERBOX.
+    };
+
+    /** @brief Creates 4-dimensional blob from image.
+     *
+     *  @details This function is an extension of @ref blobFromImage to meet more image preprocess needs.
+     *  Given input image and preprocessing parameters, and function outputs the blob.
+     *
+     *  @param image input image (all with 1-, 3- or 4-channels).
+     *  @param param struct of ImagePParams, contains all parameters needed by image preprocessing.
+     *  @return 4-dimensional Mat.
+     */
+    CV_EXPORTS_W Mat blobFromImageParam(InputArray image, const ImagePParam& param = ImagePParam());
+
+    /** @brief Creates 4-dimensional blob from series of images.
+     *  @details This is an overloaded member function, provided for convenience.
+     *           It differs from the above function only in what argument(s) it accepts.
+     */
+    CV_EXPORTS_W void blobFromImageParam(InputArray image, OutputArray blob, const ImagePParam& param = ImagePParam());
```

Comment:
Maybe we should just retain the 2 variants with OutputArray blob. Also, modify the name to
"blobFromImageWithParams", otherwise it's grammatically incorrect